File: packages/telemetry/src/__tests__/parse-response.test.ts
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { parseInstantQueryResult } from '../query/parse-response.js';

const fixturesDir = join(dirname(fileURLToPath(import.meta.url)), 'fixtures');

function loadFixture(name: string) {
	return JSON.parse(readFileSync(join(fixturesDir, name), 'utf8'));
}

describe('parseInstantQueryResult', () => {
	it('parses vector with sample value', () => {
		const result = parseInstantQueryResult(loadFixture('prometheus-vector-pass.json'));
		expect(result).toEqual({ ok: true, value: 0.02 });
	});

	it('parses scalar result type', () => {
		const result = parseInstantQueryResult(loadFixture('prometheus-scalar-pass.json'));
		expect(result).toEqual({ ok: true, value: 0.015 });
	});

	it('fails closed on empty vector', () => {
		const result = parseInstantQueryResult(loadFixture('prometheus-vector-empty.json'));
		expect(result).toEqual({ ok: false, reason: 'no_data' });
	});

	it('fails closed on NaN sample', () => {
		const result = parseInstantQueryResult(loadFixture('prometheus-nan-value.json'));
		expect(result).toEqual({ ok: false, reason: 'non_finite_value' });
	});
});

File: kotlin/modules/telemetry/src/test/kotlin/com/ffpromo/telemetry/query/ParseResponseTest.kt
package com.ffpromo.telemetry.query

import com.ffpromo.telemetry.client.PrometheusInstantQueryData
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class ParseResponseTest {
    private val json = Json { ignoreUnknownKeys = true }

    private fun loadFixture(name: String): PrometheusInstantQueryData {
        val element = json.parseToJsonElement(
            this::class.java.getResource("/fixtures/$name")!!.readText(),
        ).jsonObject
        return PrometheusInstantQueryData(
            resultType = element.getValue("resultType").jsonPrimitive.content,
            result = element.getValue("result"),
        )
    }

    @Test
    fun `parses vector with sample value`() {
        val result = parseInstantQueryResult(loadFixture("prometheus-vector-pass.json"))
        assertEquals(ParseResult.Ok(0.02), result)
    }

    @Test
    fun `parses scalar result type`() {
        val result = parseInstantQueryResult(loadFixture("prometheus-scalar-pass.json"))
        assertEquals(ParseResult.Ok(0.015), result)
    }

    @Test
    fun `fails closed on empty vector`() {
        val result = parseInstantQueryResult(loadFixture("prometheus-vector-empty.json"))
        assertEquals(ParseResult.Fail(ParseFailReason.no_data), result)
    }

    @Test
    fun `fails closed on NaN sample`() {
        val result = parseInstantQueryResult(loadFixture("prometheus-nan-value.json"))
        assertEquals(ParseResult.Fail(ParseFailReason.non_finite_value), result)
    }
}
